FeatureExtraction/Texture/texture_features.py

```diff
@@ -0,0 +1,412 @@
+    matrix[matrix < 0] = 0
+    return matrix
+
+
+def calculate_entropy(glcm: GLCM):
+    if not isinstance(glcm, GLCM):
+        raise TypeError("glcm should be an instance of GLCM class")
+    distances = glcm.shape[2]
+    angles = glcm.shape[3]
+    results = np.zeros((distances, angles))
+    for distance in range(0, distances):
+        for angle in range(0, angles):
+            log2_matrix = np.log2(glcm[:, :, distance, angle])
+            log2_matrix[np.isinf(log2_matrix)] = 0
+            result = np.sum(
+                np.multiply(glcm[:, :, distance, angle], log2_matrix))
+            results[distance, angle] = -copy(result)
+    return results
+
+
+def calculate_HXY2(glcm: GLCM):
+    if not isinstance(glcm, GLCM):
+        raise TypeError("glcm should be an instance of GLCM class")
+    y_shape, x_shape = glcm.shape[0:2]
+    distances = glcm.shape[2]
+    angles = glcm.shape[3]
+    results = np.zeros((distances, angles))
+    for distance in range(0, distances):
+        for angle in range(0, angles):
+            result = 0
+            for i in range(0, x_shape):
+                row_sum = np.sum(glcm[i, :, distance, angle])
+                for j in range(0, y_shape):
+                    column_sum = np.sum(glcm[:, j, distance, angle])
+                    row_col_product = row_sum * column_sum
+                    if row_col_product != 0:
+                        result += row_col_product * np.log(row_col_product)
+            results[distance, angle] = -copy(result)
+    return results
+
+
+def calculate_IMC2(glcm: GLCM):
+    if not isinstance(glcm, GLCM):
+        raise TypeError("glcm should be an instance of GLCM class")
+    distances = glcm.shape[2]
+    angles = glcm.shape[3]
+    entropy = calculate_entropy(glcm)
+    hxy2 = calculate_HXY2(glcm)
+    results = np.zeros((distances, angles))
+    for distance in range(0, distances):
+        for angle in range(0, angles):
+            result = sqrt(
+                1 - exp(-2. * (entropy[distance, angle] - hxy2[distance, angle])))
+            results[distance, angle] = copy(result)
+    return results
+
+
+def calculate_IDMN(glcm: GLCM):
+    if not isinstance(glcm, GLCM):
+        raise TypeError("glcm should be an instance of GLCM class")
+    y_shape, x_shape = glcm.shape[0:2]
+    x_indexes = np.linspace(0, x_shape - 1, x_shape)
+    y_indexes = np.linspace(0, y_shape - 1, y_shape)
+    cols, rows = np.meshgrid(x_indexes, y_indexes)
+    distances = glcm.shape[2]
+    angles = glcm.shape[3]
+    results = np.zeros((distances, angles))
+    for distance in range(0, distances):
+        for angle in range(0, angles):
+            single_glcm = glcm[:, :, distance, angle].reshape(x_shape, y_shape)
+            denominator = np.subtract(rows, cols)
+            denominator = np.power(denominator, 2)
+            denominator = np.divide(denominator, pow(glcm.levels, 2))
+            denominator = np.add(denominator, 1)
```

Comment:
I think all of the above operations can be packed in a function.

Reply:
I believe it's not necessary. I do not use these operations anywhere else and it's already packed into numpy functions.